File: proj.html5/Sources/Utils/List.js
/**
 * Tooflya Inc. Development
 *
 * @author Igor Mats from Tooflya Inc.
 * @copyright (c) 2013 by Igor Mats
 *
 * @license http://opensource.org/licenses/gpl-license.php GNU Public License v2
 * @all rights granted under this License are granted for the term of
 * copyright on the Program, and are irrevocable provided the stated
 * conditions are met. This License explicitly affirms your unlimited
 * permission to run the unmodified Program. The output from running a
 * covered work is covered by this License only if the output, given its
 * content, constitutes a covered work. This License acknowledges your
 * rights of fair use or other equivalent, as provided by copyright law.
 *
 * @version
 *
 */

cc.List = cc.Node.extend({
    init:function(p_Width, p_Height, p_MaxWidth, p_MaxHeight, p_ListInitialCenterX, p_ListInitialCenterY, pParent) {
        this.m_ParentType = 0;
        this.m_Type = 2;

        this.m_Width = p_Width;
        this.m_Height = p_Height;

        this.m_MaxWidth = p_MaxWidth;
        this.m_MaxHeight = p_MaxHeight;

        this.m_ListInitialCenterX = p_ListInitialCenterX;
        this.m_ListInitialCenterY = p_ListInitialCenterY;

        this.m_Background = cc.Entity.createWithFilename(s_ListBackground, this);
        this.m_Background.getTexture().setTexParameters({
            minFilter: gl.LINEAR,
            magFilter: gl.LINEAR,
            wrapS: gl.REPEAT,
            wrapT: gl.REPEAT
        });

        this.m_Background.create().setCenterPosition(cc.Options.camera.center.x, cc.Options.camera.center.y);

        var clipper = cc.ClippingNode.create();
        var stencil = cc.DrawNode.create();

        clipper.setStencil(stencil);
        clipper.addChild(this);

        stencil.setContentSize(cc.size(200, 200));

        if(pParent) {
            pParent.addChild(clipper);
        }
    },

    draw:function(ctx) {
        var gl = ctx || cc.renderContext;

        gl.enable(gl.STENCIL_TEST);

        cc.kmGLPushMatrix();
        this.transform();

        cc.kmGLPopMatrix();

        gl.disable(gl.STENCIL_TEST);
    },
    
    update:function(dt) {
        if(this.m_PostUpdate)
        {
            var x = this.getPosition().x;
            var y = this.getPosition().y;

            this.setPosition(cc.p(x, y - this.m_SpeedY));

            this.m_SpeedY -= this.m_LastDistanceY > 0 ? 0.5 : -0.5;

            if(y < 0) {
                this.runAction(cc.MoveTo.create(0.3, cc.p(x, 0)));

                this.m_PostUpdate = false;
            } else if(y > this.m_MaxHeight) {
                this.runAction(cc.MoveTo.create(0.3, cc.p(x, this.m_MaxHeight)));
            
                this.m_PostUpdate = false;
            }

            if(x > 0) {
                this.runAction(cc.MoveTo.create(0.3, cc.p(0, y)));
            
                this.m_PostUpdate = false;
            } else if(x < -this.m_MaxWidth) {
                this.runAction(cc.MoveTo.create(0.3, cc.p(this.m_MaxWidth, y)));
            
                this.m_PostUpdate = false;
            }

            if(this.m_LastDistanceY > 0 ? this.m_SpeedY <= 0.0 : this.m_SpeedY >= 0.0) {
                this.m_PostUpdate = false;
            }
        }

        this.m_Background.setTextureRect(cc.RectMake(0, this.getPosition().y, this.m_Background.getTextureRect().size.width, this.m_Height));
        this.m_Background.setCenterPosition(this.m_ListInitialCenterX + this.getPosition().x, this.m_ListInitialCenterY - this.getPosition().y);
    },

    setRegisterAsTouchable:function(p_Touchable) {
        this.m_IsRegisterAsTouchable = p_Touchable;

        if(this.m_IsRegisterAsTouchable) {
            cc.Director.getInstance().getTouchDispatcher().addTargetedDelegate(this, 0, true); // TODO: Add modal view.
        } else {
            cc.Director.getInstance().getTouchDispatcher().removeDelegate(this);
        }
    },
    isRegisterAsTouchable:function() {
        return this.m_IsRegisterAsTouchable;
    },

    containsTouchLocation:function(touch) {
        var x = touch.getLocation().x;
        var y = touch.getLocation().y;

        if(this.m_ParentType == 0) {
            return x > this.m_ListInitialCenterX - this.m_Width / 2 && x < this.m_ListInitialCenterX + this.m_Width / 2 && y < this.m_ListInitialCenterY + this.m_Height / 2 && y > this.m_ListInitialCenterY - this.m_Height / 2;
        } else if(this.m_ParentType == 1) {
            return x > cc.Options.camera.center.x- this.m_Width / 2 && x < cc.Options.camera.center.x + this.m_Width / 2 && y < cc.Options.camera.center.y + this.m_Height / 2 && y > cc.Options.camera.center.y - this.m_Height / 2/* && ((Popup*) this->getParent())->isShowed()*/;
        }
        
        return true;
    },

    onTouchBegan:function(touch, event) {
        this.stopAllActions();

        this.m_PostUpdate = false;
        
        this.m_StartCoordinateX = touch.getLocation().x;
        this.m_StartCoordinateY = touch.getLocation().y;

        this.m_StartPositionCoordinateX = this.getPosition().x;
        this.m_StartPositionCoordinateY = this.getPosition().y;
        
        return this.containsTouchLocation(touch);
    },
    onTouchMoved:function(touch, event) {
        if(this.containsTouchLocation(touch)) {
            var x, y;

            if(this.m_Type == 0) {
                x = this.m_StartPositionCoordinateX + touch.getLocation().x - this.m_StartCoordinateX;
                y = this.m_StartPositionCoordinateY + touch.getLocation().y - this.m_StartCoordinateY;
            } else if(this.m_Type == 1) {
                x = this.m_StartPositionCoordinateX + touch.getLocation().x - this.m_StartCoordinateX;
                y = this.getPosition().y;
            } else if(this.m_Type == 2) {
                x = this.getPosition().x;
                y = this.m_StartPositionCoordinateY + touch.getLocation().y - this.m_StartCoordinateY;
            }
            
            this.setPosition(cc.p(x, y));
        }
    },
    onTouchEnded:function(touch, event) {
        var x = this.getPosition().x;
        var y = this.getPosition().y;

        var willPostUpdate = true;

        if(y < 0) {
            this.runAction(cc.MoveTo.create(0.3, cc.p(x, 0)));

            willPostUpdate = false;
        } else if(y > this.m_MaxHeight) {
            this.runAction(cc.MoveTo.create(0.3, cc.p(x, this.m_MaxHeight)));

            willPostUpdate = false;
        }

        if(x > 0) {
            this.runAction(cc.MoveTo.create(0.3, cc.p(0, y)));

            willPostUpdate = false;
        } else if(x < -this.m_MaxWidth) {
            this.runAction(cc.MoveTo.create(0.3, cc.p(this.m_MaxWidth, y)));

            willPostUpdate = false;
        }

        this.m_LastDistanceX = this.m_StartPositionCoordinateX - x; // TODO: Check it!
        this.m_LastDistanceY = this.m_StartPositionCoordinateY - y;

        this.m_PostUpdate = willPostUpdate;

        this.m_SpeedY = this.m_LastDistanceY / 10.0;
    },

    onEnter:function() {
        this.scheduleUpdate();

        this._super();

        this.setRegisterAsTouchable(true);

        this.m_PostUpdate = false;
        this.stopAllActions();
    },
    onExit:function() {
        this.stopAllActions();
        this.unscheduleUpdate();

        this._super();

        this.setRegisterAsTouchable(false);
        this.setPosition(cc.p(this.getPosition().x, 0)); // TODO: Check it for horizontal list.
    }
});

cc.List.create = function(p_Width, p_Height, p_MaxWidth, p_MaxHeight, p_ListInitialCenterX, p_ListInitialCenterY, pParent) {
    var screen = new cc.List();
    screen.init(p_Width, p_Height, p_MaxWidth, p_MaxHeight, p_ListInitialCenterX, p_ListInitialCenterY, pParent);
    
    return screen;
};

File: Classes/Utils/List.cpp
#ifndef CONST_LIST
#define CONST_LIST

#include "List.h"

// ===========================================================
// Inner Classes
// ===========================================================

// ===========================================================
// Constants
// ===========================================================

// ===========================================================
// Fields
// ===========================================================

// ===========================================================
// Constructors
// ===========================================================

List::~List()
{
    this->removeAllChildrenWithCleanup(true);
}

List::List(float pWidth, float pHeight, float pMaxWidth, float pMaxHeight, float pListInitialCenterX, float pListInitialCenterY, const char* pListTextureFileName, CCNode* pParent) :
	mType(0),
	mParentType(0),
	mWidth(0),
	mHeight(0),
	mMaxWidth(0),
	mMaxHeight(0),
	mListInitialCenterX(0),
	mListInitialCenterY(0),
	mStartCoordinateY(0),
	mStartCoordinateX(0),
	mStartPositionCoordinateY(0),
	mStartPositionCoordinateX(0),
	mLastDistanceX(0),
	mLastDistanceY(0),
	mSpeedX(0),
	mSpeedY(0),
	mPostUpdate(0),
	mListSroll(0),
	mParent(0),
	mSpriteBatch(0)
	{
		this->mParent = static_cast<Entity*>(pParent);

		this->mParent->addChild(this);
    
        int size = -1;
        
        if(strcmp(pListTextureFileName, "about_scroll_fill_small@2x.png")) size = 1024; else size = 512;
		this->mListSroll = Entity::create(pListTextureFileName, 0, 0, size, size, this);
		this->mListSroll->create()->setRepeatTexture(true);

		this->mWidth = pWidth;
		this->mHeight = pHeight;

		this->mMaxWidth = pMaxWidth;
		this->mMaxHeight = pMaxHeight;

		this->mListInitialCenterX = pListInitialCenterX;
		this->mListInitialCenterY = pListInitialCenterY;
    
		this->mType = TYPE_BOTH;
		this->mParentType = PARENT_TYPE_SIMPLE;

		this->mPostUpdate = false;
	}

// ===========================================================
// Methods
// ===========================================================

void List::setListType(int pType)
{
    this->mType = pType;
}

void List::setParentType(int pType)
{
    this->mParentType = pType;
}

// ===========================================================
// Override Methods
// ===========================================================

void List::update(float pDeltaTime)
{
    CCLayer::update(pDeltaTime);

    if(this->mPostUpdate)
    {
        float x = this->getPosition().x;
        float y = this->getPosition().y;

        this->setPosition(ccp(x, y - this->mSpeedY));

        this->mSpeedY -= this->mLastDistanceY > 0 ? 0.5 : -0.5;

        if(y < 0)
        {
            this->runAction(CCMoveTo::create(0.3, ccp(x, 0)));

            this->mPostUpdate = false;
        }
        else if(y > this->mMaxHeight)
        {
            this->runAction(CCMoveTo::create(0.3, ccp(x, this->mMaxHeight)));
            
            this->mPostUpdate = false;
        }

        if(x > 0)
        {
            this->runAction(CCMoveTo::create(0.3, ccp(0, y)));
            
            this->mPostUpdate = false;
        }
        else if(x < -this->mMaxWidth)
        {
            this->runAction(CCMoveTo::create(0.3, ccp(this->mMaxWidth, y)));
            
            this->mPostUpdate = false;
        }

        if(this->mLastDistanceY > 0 ? this->mSpeedY <= 0.0 : this->mSpeedY >= 0.0)
        {
            this->mPostUpdate = false;
        }
    }

    this->mListSroll->setTextureRect(CCRectMake(0, this->getPosition().y, this->mListSroll->getTextureRect().size.width, this->mHeight));
    this->mListSroll->setCenterPosition(this->mListInitialCenterX + this->getPosition().x, this->mListInitialCenterY - this->getPosition().y);
}

bool List::ccTouchBegan(CCTouch* touch, CCEvent* event)
{
    this->stopAllActions();

    this->mPostUpdate = false;
    
    this->mStartCoordinateX = touch->getLocation().x;
    this->mStartCoordinateY = touch->getLocation().y;

    this->mStartPositionCoordinateX = this->getPosition().x;
    this->mStartPositionCoordinateY = this->getPosition().y;
    
    this->ml = touch->getLocation().x;
    
    return this->containsTouchLocation(touch);
}

void List::ccTouchMoved(CCTouch* touch, CCEvent* event)
{
    if(this->containsTouchLocation(touch))
    {
        float x, y;

        if(this->mType == TYPE_BOTH)
        {
            x = this->mStartPositionCoordinateX + touch->getLocation().x - this->mStartCoordinateX;
            y = this->mStartPositionCoordinateY + touch->getLocation().y - this->mStartCoordinateY;
        }
        else if(this->mType == TYPE_HORIZONTAL)
        {
            x = this->mStartPositionCoordinateX + touch->getLocation().x - this->mStartCoordinateX;
            y = this->getPosition().y;
        }
        else if(this->mType == TYPE_VERTICAL)
        {
            x = this->getPosition().x;
            y = this->mStartPositionCoordinateY + touch->getLocation().y - this->mStartCoordinateY;
        }
        
        this->setPosition(ccp(x, y));
        
        if(abs(this->ml - touch->getLocation().x) >= Utils::coord(10))
        {
            this->ml = touch->getLocation().x;
            
            this->mLastMoveTime = Utils::millisecondNow();
        }
    }
}

void List::ccTouchEnded(CCTouch* touch, CCEvent* event)
{
    float x = this->getPosition().x;
    float y = this->getPosition().y;

    bool kWillPostUpdate = true;

    if(y < 0)
    {
        this->runAction(CCMoveTo::create(0.3, ccp(x, 0)));

        kWillPostUpdate = false;
    }
    else if(y > this->mMaxHeight)
    {
        this->runAction(CCMoveTo::create(0.3, ccp(x, this->mMaxHeight)));

        kWillPostUpdate = false;
    }

    if(x > 0)
    {
        this->runAction(CCMoveTo::create(0.3, ccp(0, y)));

        kWillPostUpdate = false;
    }
    else if(x < -this->mMaxWidth)
    {
        this->runAction(CCMoveTo::create(0.3, ccp(this->mMaxWidth, y)));

        kWillPostUpdate = false;
    }

    this->mLastDistanceX = this->mStartPositionCoordinateX - x; // TODO: Check it!
    this->mLastDistanceY = this->mStartPositionCoordinateY - y;

    this->mPostUpdate = kWillPostUpdate;

    if(Utils::millisecondNow() - this->mLastMoveTime >= 100)
    {
        this->mPostUpdate = false;
    }

    this->mSpeedY = this->mLastDistanceY / 10.0;
}

bool List::containsTouchLocation(CCTouch* touch)
{
    float x = touch->getLocation().x;
    float y = touch->getLocation().y;

    if(this->mParentType == PARENT_TYPE_SIMPLE)
    {
        return x > this->mListInitialCenterX - this->mWidth / 2 && x < this->mListInitialCenterX + this->mWidth / 2 && y < this->mListInitialCenterY + this->mHeight / 2 && y > this->mListInitialCenterY - this->mHeight / 2;
    }
    else if(this->mParentType == PARENT_TYPE_POPUP)
    {
        return x > Options::CAMERA_CENTER_X - this->mWidth / 2 && x < Options::CAMERA_CENTER_X + this->mWidth / 2 && y < Options::CAMERA_CENTER_Y + this->mHeight / 2 && y > Options::CAMERA_CENTER_Y - this->mHeight / 2 && ((Popup*) this->getParent())->isShowed();
    }
    
    return true;
}

void List::onEnter()
{
    CCDirector* pDirector = CCDirector::sharedDirector();
    pDirector->getTouchDispatcher()->addTargetedDelegate(this, 0, false);
    
    CCLayer::onEnter();

    this->scheduleUpdate();

    this->setPosition(ccp(this->getPosition().x, 0)); // TODO: Check it for horizontal list.
}

void List::onExit()
{
    CCDirector* pDirector = CCDirector::sharedDirector();
    pDirector->getTouchDispatcher()->removeDelegate(this);
    
    CCLayer::onExit();

    this->unscheduleAllSelectors();

    this->mPostUpdate = false;
    this->stopAllActions();
}

void List::visit()
{
    kmGLPushMatrix();
    glEnable(GL_SCISSOR_TEST);

    if(this->mParentType == PARENT_TYPE_SIMPLE)
    {
        CCEGLView::sharedOpenGLView()->setScissorInPoints(this->mListInitialCenterX - this->mWidth / 2, this->mListInitialCenterY - this->mHeight / 2, this->mWidth, this->mHeight);
    }
    else if(this->mParentType == PARENT_TYPE_POPUP)
    {
        CCEGLView::sharedOpenGLView()->setScissorInPoints(Options::CAMERA_CENTER_X - (this->mWidth / 2) * this->mParent->getScaleX(), Options::CAMERA_CENTER_Y - (this->mHeight / 2) * this->mParent->getScaleY(), this->mWidth * this->mParent->getScaleX(), this->mHeight * this->mParent->getScaleY());
    }

    CCNode::visit();

    glDisable(GL_SCISSOR_TEST);
    kmGLPopMatrix();
}

void List::draw()
{
    CCLayer::draw();
}

#endif